README.md
# Minecraft-ProxyServer 
![Release](https://github.com/bangbang93/minecraft-proxy/workflows/Release/badge.svg)
![Testing](https://github.com/bangbang93/minecraft-proxy/workflows/Testing/badge.svg)

反向代理Minecraft，完整支持Forge，支持正版登录。若后端服务器是Spigot或者Cauldron，可以通过spigot.yml启用bungeecord模式来获取正确的客户端IP

## 和BungeeCord的区别
minecraft-proxy更像NGINX。
BungeeCord的目标是把多个Minecraft实例用起来像一个，minecraft-proxy的目标是让多个Minecraft实例共享25565端口，就像http应用的虚拟主机。

## 安装
### 运行环境
#### Windows/Linux/macOS
请参阅各平台<https://nodejs.org/zh-cn/>的安装方法
### 安装包
- 去<https://github.com/bangbang93/minecraft-proxy/releases>下载一个版本
- 解压
    ```shell script
    tar xvf minecraft-proxy.tar.gz
    ```
- 运行
    ```shell script
    bin/mcproxy
    ```
### Docker
```shell script
docker run -p 25565:25565 -v ./config:/opt/minecraft-proxy/config bangbang93/minecraft-proxy
```
#### 国内镜像
```shell
docker run -p 25565:25565 -v ./config:/opt/minecraft-proxy/config registry.bangbang93.com/bangbang93/minecraft-proxy
```
### 手动编译
```shell script
git clone git@github.com:bangbang93/minecraft-proxy.git
cd minecraft-proxy
npm ci
npm run build
```
#### 国内镜像
https://gitlab.bangbang93.com/bangbang93/minecraft-proxy

## 配置 config.yml
```yaml
proxy:
  port: 25565 #监听端口
  host: 0.0.0.0 #监听ip
  workers: 0 #worker数量，以便充分利用多核服务器，0代表用上全部核心
defaultServer: localhost #默认服务器，若不想指定，则删除此行
servers:
  - serverName: localhost #服务连接的主机名
    proxyHost: localhost # 后端服务实际ip
    proxyPort: 25565 #后端服务实际端口
    version: 1.8.7 #版本
    handlePing: true #是否接管ping响应
    onlineMode: true
    ping: #接管ping时必填
      maxPlayer: 100 #最大玩家
      description: motd #motd
      favicon: data:image/png;base64, #没有就不用设置
  - serverName: 192.168.193.7
    proxyHost: 192.168.193.1
    proxyPort: 25565
    version: 1.8.7
    handlePing: true
    useProxy: true # 使用PROXY协议，当后端服务是Bungee时，同样启用PROXY协议，可以正确获取客户端IP
blockList:
  ips: #同时会阻拦掉ping请求
    - 114.114.114.114
    - 10.0.0.0/32
  usernames:
    - BadGays
  uuids: # 只有开了正版验证的服务端可以封禁uuid，离线服务器只能用usernames
    - 12b8b9e0eba0ca37e935f037ff1ae90d
profileEndpoint: https://api.mojang.com/profiles/minecraft # 若使用authlib-injector，则将改地址替换为对应的api地址
```

profileEndpoint地址详见[authlib-injector文档](https://github.com/yushijinhun/authlib-injector/wiki/Yggdrasil-%E6%9C%8D%E5%8A%A1%E7%AB%AF%E6%8A%80%E6%9C%AF%E8%A7%84%E8%8C%83#%E6%8C%89%E5%90%8D%E7%A7%B0%E6%89%B9%E9%87%8F%E6%9F%A5%E8%AF%A2%E8%A7%92%E8%89%B2)

默认服务器： 当未找到对应的服务器主机名时，会连接默认服务器，若无默认服务器，则会断开连接

## 拆分配置文件
参阅 [/config](/config)

## 重载
mcproxy支持无中断重载，通过向master进程发送SIGUSR1信号即可实现

## bungeecord模式
找到spigot.yml，修改settings.bungeecord为true，重启服务器即可

## Roadmap
- [ ] 插件机制(WIP)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.5.0](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.10...v1.5.0)

- Bump minimatch from 3.0.4 to 3.1.2 [`#63`](https://github.com/bangbang93/minecraft-proxy/pull/63)
- Bump class-validator from 0.13.2 to 0.14.0 [`#62`](https://github.com/bangbang93/minecraft-proxy/pull/62)
- Bump validator and class-validator [`#60`](https://github.com/bangbang93/minecraft-proxy/pull/60)
- Bump jose from 4.4.0 to 4.9.3 [`#59`](https://github.com/bangbang93/minecraft-proxy/pull/59)
- feat: upgrade minecraft-data Resolves #67 [`#67`](https://github.com/bangbang93/minecraft-proxy/issues/67)

#### [v1.4.10](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.9...v1.4.10)

> 28 July 2022

- support 1.19 [`#58`](https://github.com/bangbang93/minecraft-proxy/issues/58)
- Release 1.4.10 [`6b28a75`](https://github.com/bangbang93/minecraft-proxy/commit/6b28a75586bc1142cff4eebefced58a042ac05af)
- 找不到有效版本时，fallback到填入的版本 [`0491f8f`](https://github.com/bangbang93/minecraft-proxy/commit/0491f8faa0e85f3a175d968bfb1118093f5a1921)

#### [v1.4.9](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.8...v1.4.9)

> 14 July 2022

- Bump moment from 2.29.2 to 2.29.4 [`#56`](https://github.com/bangbang93/minecraft-proxy/pull/56)
- 修正远程地址 [`0ca53d4`](https://github.com/bangbang93/minecraft-proxy/commit/0ca53d4c0ad323540cf1fbeedeadae219c5837cf)
- get uuid给个超时时间 [`becc5f4`](https://github.com/bangbang93/minecraft-proxy/commit/becc5f492128e849355e826528d74ca64655f698)
- ping响应后断开连接 [`d09b6f1`](https://github.com/bangbang93/minecraft-proxy/commit/d09b6f1a8d173334431179038c78b7c48b53f4fc)

#### [v1.4.8](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.8-2...v1.4.8)

> 8 July 2022

- 修正新版minecraft-data支持 Resolve #55 [`#55`](https://github.com/bangbang93/minecraft-proxy/issues/55)
- Release 1.4.8 [`464d73f`](https://github.com/bangbang93/minecraft-proxy/commit/464d73fce3c4d18bc3d3a94454bd12b282be6638)

#### [v1.4.8-2](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.8-1...v1.4.8-2)

> 6 July 2022

- Release 1.4.8-2 [`9359e8f`](https://github.com/bangbang93/minecraft-proxy/commit/9359e8fc79b48f6125a3ffe3ca0924b4f4c23bb8)
- fix pkg build [`74e1541`](https://github.com/bangbang93/minecraft-proxy/commit/74e1541ea78ed362667477fe6e03ff503a8099aa)

#### [v1.4.8-1](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.8-0...v1.4.8-1)

> 6 July 2022

- 补一下pkg的依赖 [`7d7bf92`](https://github.com/bangbang93/minecraft-proxy/commit/7d7bf92efd328e9c2571656763f056aba58b2932)
- Release 1.4.8-1 [`beb3dc8`](https://github.com/bangbang93/minecraft-proxy/commit/beb3dc8e2a490cf69a595cb20e1a785bba64390c)

#### [v1.4.8-0](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.7...v1.4.8-0)

> 6 July 2022

- Bump got from 10.4.0 to 11.8.5 [`#53`](https://github.com/bangbang93/minecraft-proxy/pull/53)
- 优化白名单机制，增加消息配置 [`#52`](https://github.com/bangbang93/minecraft-proxy/pull/52)
- Bump minimist from 1.2.5 to 1.2.6 [`#46`](https://github.com/bangbang93/minecraft-proxy/pull/46)
- Bump moment from 2.24.0 to 2.29.2 [`#49`](https://github.com/bangbang93/minecraft-proxy/pull/49)
- fix #48 编译安装启动报错 [`#48`](https://github.com/bangbang93/minecraft-proxy/issues/48)
- Release 1.4.8-0 [`b7ca10a`](https://github.com/bangbang93/minecraft-proxy/commit/b7ca10acc8d1b534c0dc9b5d816ad7e9ad8affbb)
- Create LICENSE [`ee35d34`](https://github.com/bangbang93/minecraft-proxy/commit/ee35d34b542da776e6e3494a82817c329c5cee65)
- 修改pkg打包版本 [`ab5fdd6`](https://github.com/bangbang93/minecraft-proxy/commit/ab5fdd655b50d7192473440917ec1f99ba4e4c7e)

#### [v1.4.7](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.6...v1.4.7)

> 28 March 2022

- 正确设置客户端版本 [`58f7789`](https://github.com/bangbang93/minecraft-proxy/commit/58f7789b9c13cfb9fcca185f42b3f3ed6b614c43)
- Release 1.4.7 [`5855f7e`](https://github.com/bangbang93/minecraft-proxy/commit/5855f7ed4ddb0d4cd56577a55eccc650a8e05c7d)
- 优化代码 [`6bfcc82`](https://github.com/bangbang93/minecraft-proxy/commit/6bfcc8291448593177c8cc7d52bcdba6f8ac3d76)

#### [v1.4.6](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.5...v1.4.6)

> 10 February 2022

- 使用for await替代on data [`cd2fc27`](https://github.com/bangbang93/minecraft-proxy/commit/cd2fc2747b1aaab587a73cc704fa5956809496c9)
- 优化release [`34f021c`](https://github.com/bangbang93/minecraft-proxy/commit/34f021ca0f433de8e78abf03e3722a471b0c7bad)
- ci打包二进制 [`7c83678`](https://github.com/bangbang93/minecraft-proxy/commit/7c83678c7cb9810ab7292fb2796055cd7f26cf9d)

#### [v1.4.5](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.4...v1.4.5)

> 27 January 2022

- 优化代码 [`aa82ce8`](https://github.com/bangbang93/minecraft-proxy/commit/aa82ce84b3d56d82030fb17ec4c28f14b950cfae)
- interface改为abstract class [`91f871b`](https://github.com/bangbang93/minecraft-proxy/commit/91f871b3df59186693af147a08c852e5d20c1d8b)
- Release 1.4.5 [`abd0982`](https://github.com/bangbang93/minecraft-proxy/commit/abd0982b5edd1f8503f1b7caae892ce5b972dbd3)

#### [v1.4.4](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.3...v1.4.4)

> 27 January 2022

- Release 1.4.4 [`637e283`](https://github.com/bangbang93/minecraft-proxy/commit/637e2831a6b97d67d5988fb3d086911374e7dfbc)
- fix hook [`1ae38ae`](https://github.com/bangbang93/minecraft-proxy/commit/1ae38aed03bbf1ac05032278e4b800786bf70620)

#### [v1.4.3](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.2...v1.4.3)

> 26 January 2022

- Release 1.4.3 [`9e145d4`](https://github.com/bangbang93/minecraft-proxy/commit/9e145d414b1045911c3b27459aa305b31c2447da)
- fix require default [`5af31b5`](https://github.com/bangbang93/minecraft-proxy/commit/5af31b56078dc7934d2c43615445c01c65b51c30)

#### [v1.4.2](https://github.com/bangbang93/minecraft-proxy/compare/v1.4.1...v1.4.2)

> 26 January 2022


function.js
/**
 * Created by bangbang on 15/01/27.
 */
var protocol = require('./protocol');
var protocolVersion = require('./protocolVersion');
var fs = require('fs');

var status = exports.status = function status(serverInfo, client){
    var connectPool = global.connections[serverInfo.host + ':' + (serverInfo.port||25565)] || {
        count: 0
    };
    var response = {
        'version':{
            'name': serverInfo['handlePing']['version'],
            'protocol': protocolVersion(serverInfo['handlePing']['version'])
        },
        'players':{
            'max': serverInfo['handlePing']['maxPlayers'],
            'online': connectPool.count
        },
        'description':{
            'text': serverInfo['handlePing']['description']
        }
    };
    if (serverInfo['handlePing']['favicon']){
        fs.exists(serverInfo['handlePing']['favicon'], function (exists){
            if (exists){
                fs.readFile(serverInfo['handlePing']['favicon'], {encoding: null}, function (err, data){
                    if (!err){
                        response.favicon = 'data:image/png;base64,' + data.toString('base64');
                    }
                    client.write(protocol.createPacketBuffer(0x00, 'status', {
                        response: JSON.stringify(response)
                    }, true));
                })
            } else {
                client.write(protocol.createPacketBuffer(0x00, 'status', {
                    response: JSON.stringify(response)
                }, true));
            }
        })
    } else {
        client.write(protocol.createPacketBuffer(0x00, 'status', {
            response: JSON.stringify(response)
        }, true));
    }
};
